### PR TITLE
PSBT version enum

### PR DESCRIPTION
**Pull request by dr-orlovsky**
This is first of the PRs to implement PSBTv2 as was discussed in #726.

I have started with converting PSBT version into an enum object, rejecting PSBTs which are not v0 or v2 (as per standard).

I have renamed version field since with v2 there will be another `tx_version` field, to avoid naming ambiguity.

I will be splitting the work on PSBTv2 into multiple PRs to have ability to discuss design decisions one-by-one - I assume many of them may be non-trivial and controversial.

### CODE REVIEW

File: src/util/psbt/mod.rs [hunk 1]
```diff
@@ -44,15 +45,51 @@ pub use self::map::{Map, Input, Output, TapTree};
...
 use util::bip32::{ExtendedPubKey, KeySource};
 
+/// Future version of PSBT which can't be parsed by this library
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct FutureVersionError(u32);
+
+impl fmt::Display for FutureVersionError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("Unknown future PSBT version ")?;
+        fmt::Display::fmt(&self.0, f)
```
[the remark this answers is not in the document]
Comment (tcharding) [reply src/util/psbt/mod.rs:55]: I tried to understand this comment but could not, what unexpected things do you mean?
```
        let e = FutureVersionError(123);
        println!("{number:0>width$}", number=123, width=6);
        println!("{number:0>width$}", number=e, width=6);
```
Produces

000123
Unknown future PSBT version 000123

[the remark this answers is not in the document]
Comment (Kixunil) [reply src/util/psbt/mod.rs:55]: Try padding with spaces. :)

**Comment by apoelstra** [comment]
We definitely should not reject parsing PSBT versions that we don't understand. PSBT is deliberately designed to allow adding extra fields that old parsers don't understand, but are able to preserve and combine.

**Comment by DanGould** [comment]
In achow's [implementation](https://github.com/bitcoin/bitcoin/blob/e661c725e18967a0e8ce94be1d64b2b2ce974116/src/rpc/rawtransaction.cpp) the conditionals handle v==0, v==1 (which errors), and v>=2 according to bip370. It makes sense to catch all v>=2 rather than error to me

**Comment by DanGould** [comment]
I was wrong. According to [the bip 174 spec](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#version-numbers) version numbers higher than we understand necessarily contain breaking changes and must be rejected.

> The Version number field exists only as a safeguard in the event that a backwards incompatible change is introduced to PSBT. **If a parser encounters a version number it does not recognize, it should exit immediately** as this indicates that the PSBT will contain types that it does not know about and cannot be ignored.

**Comment by dr-orlovsky** [comment]
Closing: do not plan to work on this PR or PSBTv2 here any further; will complete PSBTv2 and PSBTv1 re-implementation with better API in an independent psbt crate https://docs.rs/psbt/0.7.0-beta.1/psbt/index.html

File: src/util/psbt/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DanGould) [reply src/util/psbt/mod.rs:55]: What does padding with spaces do? It does what I'd expect 🤔

```rs
let e = FutureVersionError(123);
println!("{number:0>width$}", number=123, width=6);
println!("{number:0>width$}", number=e, width=6);
println!("{number: >width$}", number=e, width=6);
```
produces 
```
000123
Unknown future PSBT version 000123
Unknown future PSBT version    123
```

**Pull request closed without merging**